lib/tree.rb

```diff
@@ -16,51 +16,154 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # recursive helper method 
+  def add_helper(current_node, key, value)
+    #  base case 
+    return TreeNode.new(key, value) if current_node.nil?
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else 
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+    return current_node
+    
+  end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    # raise NotImplementedError
+    current = @root
+    return nil if current.nil?
+    return find_helper(root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current_node, key)
+    return current_node.value if current_node.key == key
+
+    if key <= current_node.key
+      current_node.left = find_helper(current_node.left, key)
+    else
+      current_node.right = find_helper(current_node.right, key)
+    end
+
+  end
+
+
+
+  def inorder_helper(current_node, list)
+    return list if current_node.nil? 
+
+    inorder_helper(current_node.left, list)
+    list << { key: current_node.key, value: current_node.value }
+    inorder_helper(current_node.right, list)
+    return list 
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
+  # Time Complexity: O(n) call visit on each node exactly once
   # Space Complexity: 
+  # current node, left subtree, right subtree
+    # 1. visit node
+    # 2. traverse left subtree
+    # 3. traverse right subtree
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
+
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(current_node, list)
+    return list if current_node.nil?
+    list <<  { key: current_node.key, value: current_node.value}
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+  end
+
+  # Time Complexity: O(n), one call made per visit on each node
+  # Space Complexity: O(n) - where n refers to the stack
+  # left right current
   def postorder
-    raise NotImplementedError
+    # raise NotImplementedError
+    return postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(current_node, list)
+    return list if current_node.nil?  
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list << { key: current_node.key, value: current_node.value}
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
   def height
```

Comment:
Space complexity is off, it's always returning the `height_helper(current_node.right, height + 1)`